File: LeetCode-Java/Letter Combinations of a Phone Number.java
public class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<String>();
        if(digits == null)
            return result;
        if(digits.length() == 0)
        {
            result.add("");
            return result;
        }
        String[] source = {"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        char[] record = new char[digits.length()];
        letterCombinationsPart(source,digits,0,record,result);
        return result;
    }
    
    void letterCombinationsPart(String[] source,String digits, int index, char[] record, List<String> result)
    {
        int numberIndex = Character.compare(digits.charAt(index),'0');
        String choice = source[numberIndex];
        if(choice.length() <= 0)
            return;
        for(int i = 0;i < choice.length(); i ++)
        {
            char c = choice.charAt(i);
            record[index] = c;
            if(index == digits.length() - 1)
            {
                String tempResult = String.valueOf(record);
                result.add(tempResult);
            }
            else
            {
                letterCombinationsPart(source,digits,index + 1, record,result);
            }
        }
        return;
    }
}

File: Leetcode-cpp/letter_combinations_of_a_phone_number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> result;
        if (digits.length() == 0) return result;
        vector<char> record;
        search(result, record, digits, 0);
        return result;
    }
    void search(vector<string>& result, vector<char>& record, const string& digits, int index) {
        if (index == digits.length()) {
            result.push_back(string(record.begin(), record.end()));
            return;
        }
        char low [] = {'a', 'd', 'g', 'j', 'm', 'p', 't', 'w'};
        char high [] = {'c', 'f', 'i', 'l', 'o', 's', 'v', 'z'};
        if (digits[index] >= '2' && digits[index] <= '9') {
            for (char c = low[digits[index] - '2']; c <= high[digits[index] - '2']; c++) {
                record.push_back(c);
                search(result, record, digits, index + 1);
                record.erase(record.end() - 1);
            }
        }
    }
};